File: server/game.go
package main

const (
	EMPTY = 0
	P1    = 1
	P2    = 2
)

type Board struct {
	Cells    [][]int `json:"cells"`
	Size     int     `json:"size"`
	WinLen   int     `json:"winLen"`
}

func NewBoard(size, winLen int) *Board {
	cells := make([][]int, size)
	for i := range cells {
		cells[i] = make([]int, size)
	}
	return &Board{Cells: cells, Size: size, WinLen: winLen}
}

func (b *Board) Clone() *Board {
	cells := make([][]int, b.Size)
	for i := range cells {
		cells[i] = make([]int, b.Size)
		copy(cells[i], b.Cells[i])
	}
	return &Board{Cells: cells, Size: b.Size, WinLen: b.WinLen}
}

func (b *Board) IsEmpty(r, c int) bool {
	return r >= 0 && r < b.Size && c >= 0 && c < b.Size && b.Cells[r][c] == EMPTY
}

func (b *Board) Place(r, c, player int) bool {
	if !b.IsEmpty(r, c) {
		return false
	}
	b.Cells[r][c] = player
	return true
}

func (b *Board) CheckWin(player int) bool {
	dirs := [][2]int{{0, 1}, {1, 0}, {1, 1}, {1, -1}}
	for r := 0; r < b.Size; r++ {
		for c := 0; c < b.Size; c++ {
			if b.Cells[r][c] != player {
				continue
			}
			for _, d := range dirs {
				dr, dc := d[0], d[1]
				ok := true
				for i := 1; i < b.WinLen; i++ {
					nr, nc := r+dr*i, c+dc*i
					if nr < 0 || nr >= b.Size || nc < 0 || nc >= b.Size || b.Cells[nr][nc] != player {
						ok = false
						break
					}
				}
				if ok {
					return true
				}
			}
		}
	}
	return false
}

func (b *Board) IsFull() bool {
	for r := 0; r < b.Size; r++ {
		for c := 0; c < b.Size; c++ {
			if b.Cells[r][c] == EMPTY {
				return false
			}
		}
	}
	return true
}

func (b *Board) AvailableMoves() [][2]int {
	var moves [][2]int
	for r := 0; r < b.Size; r++ {
		for c := 0; c < b.Size; c++ {
			if b.Cells[r][c] == EMPTY {
				moves = append(moves, [2]int{r, c})
			}
		}
	}
	return moves
}

func evaluateLine(board *Board, row, col, dr, dc, player int) int {
	score := 0
	for start := -(board.WinLen - 1); start <= 0; start++ {
		pieces := 0
		blocked := false
		for i := 0; i < board.WinLen; i++ {
			r := row + (start+i)*dr
			c := col + (start+i)*dc
			if r < 0 || r >= board.Size || c < 0 || c >= board.Size {
				blocked = true
				break
			}
			if board.Cells[r][c] == EMPTY {
				continue
			}
			if board.Cells[r][c] == player {
				pieces++
			} else {
				blocked = true
				break
			}
		}
		if !blocked && pieces > 0 {
			var val int = 1
			for i := 0; i < pieces; i++ {
				val *= 10
			}
			score += val
		}
	}
	return score
}

func AImove(board *Board, ai, human int) [2]int {
	center := board.Size / 2
	best := -1
	var bestCell [2]int
	dirs := [][2]int{{0, 1}, {1, 0}, {1, 1}, {1, -1}}

	for r := 0; r < board.Size; r++ {
		for c := 0; c < board.Size; c++ {
			if board.Cells[r][c] != EMPTY {
				continue
			}
			score := 0
			for _, d := range dirs {
				score += evaluateLine(board, r, c, d[0], d[1], ai) * 11 / 10
				score += evaluateLine(board, r, c, d[0], d[1], human)
			}
			dist := abs(r-center) + abs(c-center)
			score += (board.Size - dist) * 5 / 10
			if score > best {
				best = score
				bestCell = [2]int{r, c}
			}
		}
	}
	return bestCell
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: frontend/src/game.js
export const EMPTY = 0
export const P1 = 1
export const P2 = 2

export function createBoard(size) {
  return Array.from({ length: size }, () => Array(size).fill(EMPTY))
}

export function checkWin(board, player, winLen) {
  const n = board.length
  const dirs = [[0, 1], [1, 0], [1, 1], [1, -1]]
  for (let r = 0; r < n; r++) {
    for (let c = 0; c < n; c++) {
      if (board[r][c] !== player) continue
      for (const [dr, dc] of dirs) {
        const cells = [[r, c]]
        let ok = true
        for (let i = 1; i < winLen; i++) {
          const nr = r + dr * i
          const nc = c + dc * i
          if (nr < 0 || nr >= n || nc < 0 || nc >= n || board[nr][nc] !== player) {
            ok = false
            break
          }
          cells.push([nr, nc])
        }
        if (ok) return cells
      }
    }
  }
  return null
}

export function isFull(board) {
  return board.every(row => row.every(c => c !== EMPTY))
}

function evaluateLine(board, row, col, dr, dc, player, winLen) {
  const n = board.length
  let score = 0
  for (let start = -(winLen - 1); start <= 0; start++) {
    let pieces = 0
    let blocked = false
    for (let i = 0; i < winLen; i++) {
      const r = row + (start + i) * dr
      const c = col + (start + i) * dc
      if (r < 0 || r >= n || c < 0 || c >= n) { blocked = true; break }
      if (board[r][c] === EMPTY) continue
      if (board[r][c] === player) pieces++
      else { blocked = true; break }
    }
    if (!blocked && pieces > 0) {
      score += Math.pow(10, pieces)
    }
  }
  return score
}

export function aiMove(board, ai, human, winLen) {
  const n = board.length
  const center = Math.floor(n / 2)
  let best = -1
  let bestCell = null
  const dirs = [[0, 1], [1, 0], [1, 1], [1, -1]]
  for (let r = 0; r < n; r++) {
    for (let c = 0; c < n; c++) {
      if (board[r][c] !== EMPTY) continue
      let score = 0
      for (const [dr, dc] of dirs) {
        score += evaluateLine(board, r, c, dr, dc, ai, winLen) * 1.1
        score += evaluateLine(board, r, c, dr, dc, human, winLen)
      }
      const dist = Math.abs(r - center) + Math.abs(c - center)
      score += (n - dist) * 0.5
      if (score > best) { best = score; bestCell = [r, c] }
    }
  }
  return bestCell
}
